### PR TITLE
handle conflict with checkUnusedConstraints

### PR DESCRIPTION
**Pull request by esword**
## Before this PR
Many repos that use the shadow plugin get errors running the checkUnusedConstraints task.  The root cause is:

```
Caused by: org.gradle.api.InvalidUserDataException: Cannot change hierarchy of dependency configuration ':unshaded' after it has been resolved.
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.preventIllegalMutation(DefaultConfiguration.java:1469)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.validateMutation(DefaultConfiguration.java:1414)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.extendsFrom(DefaultConfiguration.java:404)
	at com.palantir.gradle.shadowjar.ShadowJarPlugin.lambda$setupShadowJarToShadeTheCorrectDependencies$9(ShadowJarPlugin.java:147)
	at com.palantir.gradle.shadowjar.ShadowJarPlugin.lambda$setupShadowJarToShadeTheCorrectDependencies$10(ShadowJarPlugin.java:146)
```
 
The work around has been to disabled the checkUnusedConstraints task.

## After this PR
Tell the consistent-versions plugin to exclude the shaded configuration when running that task.  The excludeConfigurations setting is only used by the checkUnusedConstraints task so this does not have any impact on other functionality.

==COMMIT_MSG==
exclude shaded configuration from checkUnusedConstraints task
==COMMIT_MSG==

### CODE REVIEW

File: src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java [hunk 1]
```diff
@@ -111,6 +112,13 @@ private void setupShadowJarToShadeTheCorrectDependencies(
             conf.setCanBeResolved(false);
         });
 
+        // Tell the consistent versions plugin not to look at the unshaded configuration for the
+        // checkUnusedConstraints task,  It forces resolution before the configuration below is complete which causes
+        // an error under gradle8.
```
Comment (CRogers) [inline src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: I think this plugin requires GCV to work properly - will need to think whether this actually keeps it working?

File: src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java [hunk 2]
```diff
@@ -111,6 +112,13 @@ private void setupShadowJarToShadeTheCorrectDependencies(
...
+        VersionRecommendationsExtension recExt =
+                project.getExtensions().getByType(VersionRecommendationsExtension.class);
+        recExt.excludeConfigurations("unshaded");
```
Comment (CRogers) [inline src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: Isn't this what we're doing to `rejectedFromShading` below (calling `ShadowJarVersionLock#excludeConfigurationFromVersionsPropsInjection`?)

File: src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java (hunk 1 above)
Comment (CRogers) [reply src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: We lock unshaded below... not sure what that means without version props injection. Perhaps it works out fine as we still get the constraints from elsewhere as it locked?

File: src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java (hunk 2 above)
Comment (CRogers) [inline src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: This direct java object manipulation also doesn't work when people are using `plugins` blocks due to different classloaders (quite common for repos to use `plugins` with GCV for some reason). Hence using groovy in `ShadowJarVersionLock#excludeConfigurationFromVersionsPropsInjection`.

Comment (esword) [reply src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: heh - ya.  I did it the "non-groovy" way to experiment with something and then didn't change it back.

File: src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java (hunk 1 above)
Comment (esword) [reply src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: My reasoning is this - The unshaded configuration is setup automatically based on the inputs of other configurations.  Therefore the dependencies will be resolved as part of those other configurations.  Similarly, the unshaded configuration is not going to have unused constraints that are not already resolved within those other configurations.

Comment (CRogers) [reply src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: I think that's enough to convince me, just need to use `excludeConfigurationFromVersionsPropsInjection` and we can merge this in

File: src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java [hunk 3]
```diff
@@ -132,7 +132,12 @@ private void setupShadowJarToShadeTheCorrectDependencies(
         // from another source, so this *should* be ok (there is a test for this).
         ShadowJarVersionLock.excludeConfigurationFromVersionsPropsInjection(project, rejectedFromShading);
 
-        unshaded.getIncoming().beforeResolve(_ignored -> {
+        unshaded.getIncoming().beforeResolve(incoming -> {
+            // do not attempt to process if this is the copy of the configuration - it means that GCV has already
+            // resolved things.
```
Comment (CRogers) [inline src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: So here, we're adding a `beforeResolve` on `unshaded`, but then the GCV copies it and that also copies the `beforeResolve` we've set? And then it's that _copied_ `beforeResolve` that tries to illegally modify the original `unshaded` configuration?

Comment (CRogers) [reply src/main/groovy/com/palantir/gradle/shadowjar/ShadowJarPlugin.java]: If this is correct, maybe just expand the comment a bit to cover the copied beforeResolves/what I said above.

File: README.md [hunk 4]
```diff
@@ -28,7 +28,7 @@ To apply this plugin, *build.gradle* should look something like:
  }
 ```
 
-*Requires [`gradle-consistent-versions`](https://github.com/palantir/gradle-consistent-versions) and Gradle 6 to work.*
+*Requires [`gradle-consistent-versions`](https://github.com/palantir/gradle-consistent-versions) and Gradle 8 to work.*
```
Comment (CRogers) [inline README.md:31]: Does this require Gradle 8? We still need to support Gradle 7 during migration.

Comment (esword) [reply README.md:31]: I thought the underlying shadow plugin required gradle 8?  Wasn't that why we had to bump this plugin's major version?

Comment (CRogers) [reply README.md:31]: Ah, good point.

**Review by CRogers — APPROVED** [review]

**Pull request merged**